hdf/src/atom.c

```diff
@@ -212,17 +292,50 @@ HAregister_atom(group_t grp,   /* IN: Group to register the object in */
      HAatom_object - Returns to the object ptr for the atom
 
  DESCRIPTION
-    Retrieves the object ptr which is associated with the atom.
+    Retrieves the object ptr which is associated with the atom. This call
+    first checks the atom cache and reverts to the general lookup if that
+    fails.
 
  RETURNS
     Returns object ptr if successful and NULL otherwise
 *******************************************************************************/
-VOIDP
-HAPatom_object(atom_t atm /* IN: Atom to retrieve object for */
+void *
+HAatom_object(atom_t atm)
+{
+    /* This unmaintainable mess was converted from a public macro in atom.h and
+     * should be cleaned up and merged with HAIatom_object()
+     */
+
+    /* Note! This is hardwired to the atom cache size being 4 */
+
+    /* Checks if the atom is in the cache, moving items around as it goes. If
+     * the atom is not found, it does a general lookup via HAIatom_object()
+     * and returns that.
+     */
+    return (atom_id_cache[0] == atm   ? atom_obj_cache[0]
+            : atom_id_cache[1] == atm ? (SWAP_CACHE(0, 1), atom_obj_cache[0])
+            : atom_id_cache[2] == atm ? (SWAP_CACHE(1, 2), atom_obj_cache[1])
+            : atom_id_cache[3] == atm ? (SWAP_CACHE(2, 3), atom_obj_cache[2])
+                                      : HAIatom_object(atm));
+}
```

Comment:
The above function is a bit ugly, but I didn't want to risk bugs by screwing with the macro code. I just wanted to drag it inside the atom.c file so the atom cache wasn't exposed via atom.h.